doc/reference/trd-uart.md

```diff
@@ -0,0 +1,527 @@
+direction of communication, but limits its speed as clock drift
+between the two sides can cause bits to be read incorrectly.
+
+The UART HIL is in the kernel crate, in module `hil::uart`. It provides five
+main traits:
+
+  * `kernel::hil::uart::Configuration`: allows a client to query how a
+    UART is configured.
+  * `kernel::hil::uart::Configure`: allows a client to configure a UART,
+    setting its speed, data width, parity, and stop bit configuration.
+  * `kernel::hil::uart::Transmit`: is for transmitting data.
+  * `kernel::hil::uart::TransmitClient`: is for handling callbacks
+    when a data transmission is complete.
+  * `kernel::hil::uart::Receive`: is for receiving data.
+  * `kernel::hil::time::ReceiveClient`: handles callbacks when data is
+    received.
+
+There are also collections of traits that combine these into more
+complete abstractions. For example, the `Uart` trait represents a
+complete UART, extending `Transmit`, `Receive`, and `Configure`.
+
+To provide a level of minimal platform independence, a port of Tock to
+a given microcontoller is expected to implement certain instances of
+these traits. This allows, for example, debug output and panic dumps
+to work across chips and platforms.
+
+This document describes these traits, their semantics, and the
+instances that a Tock chip is expected to implement. It also describes
+how the `virtual_uart` capsule allows multiple clients to share a
+UART.  This document assumes familiarity with serial ports and their
+framing: [Wikipedia's article on asynchronous serial
+communication](https://en.wikipedia.org/wiki/Asynchronous_serial_communication)
+is a good reference.
+
+2 `Configuration` and `Configure`
+===============================
+
+The `Configuration` trait allows a client to query how a UART is
+configured. The `Configure` trait allows a client to configure a UART,
+by setting is baud date, data width, parity, stop bits, and whether
+hardware flow control is enabled.
+
+These two traits are separate because there are cases when clients
+need to know the configuration but cannot set it. For example, when a UART
+is virtualized across multiple clients (e.g., so multiple sources can
+write to the console), individual clients may want to check the baud rate.
+However, they cannot set the baud rate, because that is fixed and shared
+across all of them. Similarly, some services may need to be able to set
+the UART configuration but do not need to check it.
+
+Most devices using serial ports today use 8-bit data, but some older
+devices use more or fewer bits, and hardware implementations support
+this. If the data width of a UART is set to less than 8 bits, data is
+still partitioned into bytes, and the UART sends the least significant
+bits of each byte. Suppose a UART is configured to send 7-bit
+words. If a client sends 5 bytes, the UART will send 35 bits,
+transmitting the bottom 7 bits of each byte. The most significant bit
+of each byte is ignored.
+
```

Comment:
```suggestion

To avoid ambiguity with in-memory bytes, we will refer to the 5- to 9-bit entities
sent on the wire as `characters`.
```

This change will require changing "data word" out in many places in this document,
but would bring the document in line with the terminology used in the Wikipedia article
linked at the beginning of the text as well as the datasheets I skimmed.